src/main/java/com/google/api/codegen/transformer/php/PhpModelTypeNameConverter.java

```diff
@@ -153,67 +152,48 @@ public TypeName getTypeName(ProtoElement elem) {
-        throw new IllegalStateException(
-            "Cannot determine type for cyclic message: " + parentMessage);
-      }
-      if (maxDepth == 0) {
-        throw new IllegalStateException("Cannot determine type for deeply nested message");
-      }
-
-      String parentFullName = getTypeName(parent, maxDepth - 1).getFullName();
-      String fullName = String.format("%s_%s", parentFullName, elem.getSimpleName());
-      String nickName = fullName.substring(fullName.lastIndexOf("\\") + 1);
-
-      return new TypeName(fullName, nickName);
-    }
-    return typeNameConverter.getTypeName(getTypeNameString(elem));
-  }
-
-  /**
-   * This function determines the type name as follows: If the proto type name is in TYPE_NAME_MAP,
-   * return that value. Else, split on ".", prepend '\' and capitalize each component of the
-   * namespace except the message name
-   */
-  private static String getTypeNameString(ProtoElement elem) {
     String fullName = elem.getFullName();
     if (TYPE_NAME_MAP.containsKey(fullName)) {
-      return TYPE_NAME_MAP.get(fullName);
+      return typeNameConverter.getTypeName(TYPE_NAME_MAP.get(fullName));
     }
-    String[] components = fullName.split("\\.");
-    String shortName = components[components.length - 1];
 
     StringBuilder builder = new StringBuilder();
+    LinkedList<String> components = new LinkedList<>();
+
+    while (elem != null && !(elem instanceof ProtoFile)) {
+      components.addFirst(elem.getSimpleName());
+      elem = elem.getParent();
+    }
 
-    ProtoElement parentElem = elem.getParent();
-    if (parentElem != null && parentElem instanceof ProtoFile) {
-      ProtoFile protoFile = (ProtoFile) parentElem;
+    // Create the type name based on protobuf PHP namespace
+    if (elem != null && elem instanceof ProtoFile) {
+      ProtoFile protoFile = (ProtoFile) elem;
       String namespace = protoFile.getProto().getOptions().getPhpNamespace();
-      if (Strings.isNullOrEmpty(namespace)) {
-        for (int index = 0; index < components.length - 1; index++) {
+      if (!Strings.isNullOrEmpty(namespace)) {
+        builder.append('\\').append(CharMatcher.is('\\').trimFrom(namespace));
+        for (String component : components) {
           builder
-              .append('\\')
-              .append(components[index].substring(0, 1).toUpperCase())
-              .append(components[index].substring(1));
+              .append("\\")
+              .append(component.substring(0, 1).toUpperCase())
+              .append(component.substring(1));
         }
-      } else {
-        builder.append('\\').append(CharMatcher.is('\\').trimFrom(namespace));
+        return typeNameConverter.getTypeName(builder.toString());
       }
     }
 
-    builder.append('\\').append(shortName);
-    return builder.toString();
+    // Create the type name based on the element's full name
+    for (String component : fullName.split("\\.")) {
+      builder
+          .append("\\")
```

Comment:
This step of splitting, appending `\`, capitalizing, and rejoining is duplicated. Consider the following refactor: create a helper function:

```java
private static String makeNamespacePiece(String piece) {
  return "\\" + piece.substring(0, 1).toUpperCase() + piece.substring(1);
}
```

Then use streams:

```java
String s = fullName.split("\\.").stream().transform(makeNamespacePiece).joinStrings();
```

(I don't think joinStrings() is a thing, but use a collector to join the strings in some way).